File: app/src/androidTest/java/com/door43/translationstudio/ui/newlanguage/QuestionnaireActivityUiTest.java
package com.door43.translationstudio.ui.newlanguage;

import android.content.Intent;

import com.door43.translationstudio.R;

import org.junit.Test;


import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.pressBack;
import static androidx.test.espresso.assertion.ViewAssertions.doesNotExist;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

import androidx.test.filters.LargeTest;

@LargeTest
public class QuestionnaireActivityUiTest extends NewLanguageActivityUiUtils {

    @Test
    public void fillPageBoolean() throws Exception {

        //given
        int pageNum = 0;
        boolean hideKeyboard = false;
        boolean requiredOnly = false;
        boolean valueForBooleans = false;
        boolean doNext = true;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);
        fillPage(pageNum, true, false, false, hideKeyboard);
        verifyPageLayout(pageCount(), pageNum + 1);

        //when
        fillPage(pageNum + 1, doNext, requiredOnly, valueForBooleans, hideKeyboard);

        //then
        verifyPageLayout(pageCount(), pageNum + 2);
    }


    @Test
    public void fillPageNotRequired() throws Exception {

        //given
        int pageNum = 0;
        boolean hideKeyboard = false;
        boolean requiredOnly = false;
        boolean valueForBooleans = false;
        boolean doNext = true;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);

        //when
        fillPage(pageNum, doNext, requiredOnly, valueForBooleans, hideKeyboard);

        //then
        int pageNumExpected = 1;
        verifyPageLayout(pageCount(), pageNumExpected);
    }


    @Test
    public void fillAllPagesNotRequired() throws Exception {

        //given
        int pageNum = 0;
        int fillToPage = pageCount();
        boolean hideKeyboard = false;
        boolean requiredOnly = false;
        boolean valueForBooleans = true;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);

        //when
        fillUpToPage(fillToPage, hideKeyboard, requiredOnly, valueForBooleans, false);

        //then
        verifyPageLayout(pageCount(), fillToPage - 1);
    }

    @Test
    public void fillAllPagesRequiredOnly() throws Exception {

        //given
        int pageNum = 0;
        int fillToPage = pageCount();
        boolean hideKeyboard = false;
        boolean requiredOnly = true;
        boolean valueForBooleans = true;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);

        //when
        fillUpToPage(fillToPage, hideKeyboard, requiredOnly, valueForBooleans, false);

        //then
        verifyPageLayout(pageCount(), fillToPage - 1);
    }

    @Test
    public void requiredAnswerContinue() throws Exception {

        //given
        int pageNum = 0;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);

        //when
        onView(withId(R.id.next_button)).perform(click());
        thenShouldHaveRequiredAnswerDialog();
        onView(withText(R.string.dismiss)).perform(click());

        //then
        verifyPageLayout(pageCount(), pageNum);
    }

    @Test
    public void missingAnswerContinue() throws Exception {

        //given
        int pageNum = 0;
        boolean hideKeyboard = true;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);

        //when
        onView(withId(R.id.next_button)).perform(click());
        onView(withText(R.string.missing_question_answer)).check(matches(isDisplayed())); // dialog displayed
        onView(withText(R.string.missing_question_answer)).perform(pressBack()); // dismiss
        onView(withText(R.string.missing_question_answer)).check(doesNotExist()); // dialog dismissed

        //then
        int pageNumExpected = 0;
        verifyPageLayout(pageCount(), pageNumExpected);
    }

    @Test
    public void missingAnswerCancel() throws Exception {

        //given
        int pageNum = 0;
        boolean hideKeyboard = true;
        mActivityRule.launchActivity(new Intent());
        verifyPageLayout(pageCount(), pageNum);

        //when
        onView(withId(R.id.next_button)).perform(click());
        onView(withText(R.string.missing_question_answer)).check(matches(isDisplayed())); // dialog displayed
        onView(withText(R.string.missing_question_answer)).perform(pressBack()); // dismiss dialog
        onView(withText(R.string.missing_question_answer)).check(doesNotExist()); // dialog dismissed

        //then
        verifyPageLayout(pageCount(), pageNum);
    }

}

File: app/src/androidTest/java/com/door43/translationstudio/ui/newlanguage/QuestionnaireActivityUiTest.kt
package com.door43.translationstudio.ui.newlanguage

import android.content.Intent
import androidx.test.core.app.ActivityScenario
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import com.door43.translationstudio.R
import dagger.hilt.android.testing.HiltAndroidTest
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
@LargeTest
@HiltAndroidTest
class QuestionnaireActivityUiTest : NewLanguageActivityUiUtils() {

    @Test
    @Throws(Exception::class)
    fun fillPageBoolean() {
        //given

        val pageNum = 0
        val hideKeyboard = false
        val requiredOnly = false
        val valueForBooleans = false
        val doNext = true

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)
        fillPage(pageNum, doNext = true, requiredOnly = false, valueForBooleans = false, hideKeyboard)
        verifyPageLayout(pageCount(), pageNum + 1)

        //when
        fillPage(pageNum + 1, doNext, requiredOnly, valueForBooleans, hideKeyboard)

        //then
        verifyPageLayout(pageCount(), pageNum + 2)
    }


    @Test
    @Throws(Exception::class)
    fun fillPageNotRequired() {
        //given

        val pageNum = 0
        val hideKeyboard = false
        val requiredOnly = false
        val valueForBooleans = false
        val doNext = true

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)

        //when
        fillPage(pageNum, doNext, requiredOnly, valueForBooleans, hideKeyboard)

        //then
        val pageNumExpected = 1
        verifyPageLayout(pageCount(), pageNumExpected)
    }


    @Test
    @Throws(Exception::class)
    fun fillAllPagesNotRequired() {
        //given

        val pageNum = 0
        val fillToPage: Int = pageCount()
        val hideKeyboard = false
        val requiredOnly = false
        val valueForBooleans = true

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)

        //when
        fillUpToPage(fillToPage, hideKeyboard, requiredOnly, valueForBooleans, false)

        //then
        verifyPageLayout(pageCount(), fillToPage - 1)
    }

    @Test
    @Throws(Exception::class)
    fun fillAllPagesRequiredOnly() {
        //given

        val pageNum = 0
        val fillToPage: Int = pageCount()
        val hideKeyboard = false
        val requiredOnly = true
        val valueForBooleans = true

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)

        //when
        fillUpToPage(fillToPage, hideKeyboard, requiredOnly, valueForBooleans, false)

        //then
        verifyPageLayout(pageCount(), fillToPage - 1)
    }

    @Test
    @Throws(Exception::class)
    fun requiredAnswerContinue() {
        //given

        val pageNum = 0

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)

        //when
        Espresso.onView(ViewMatchers.withId(R.id.next_button)).perform(ViewActions.click())
        thenShouldHaveRequiredAnswerDialog()
        Espresso.onView(ViewMatchers.withText(R.string.dismiss)).perform(ViewActions.click())

        //then
        verifyPageLayout(pageCount(), pageNum)
    }

    @Test
    @Throws(Exception::class)
    fun missingAnswerContinue() {
        //given

        val pageNum = 0
        val hideKeyboard = true

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)

        //when
        Espresso.onView(ViewMatchers.withId(R.id.next_button)).perform(ViewActions.click())
        Espresso.onView(ViewMatchers.withText(R.string.missing_question_answer))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed())) // dialog displayed
        Espresso.onView(ViewMatchers.withText(R.string.missing_question_answer))
            .perform(ViewActions.pressBack()) // dismiss
        Espresso.onView(ViewMatchers.withText(R.string.missing_question_answer))
            .check(ViewAssertions.doesNotExist()) // dialog dismissed

        //then
        val pageNumExpected = 0
        verifyPageLayout(pageCount(), pageNumExpected)
    }

    @Test
    @Throws(Exception::class)
    fun missingAnswerCancel() {
        //given

        val pageNum = 0
        val hideKeyboard = true

        val intent = Intent(appContext, NewTempLanguageActivity::class.java)
        val scenario = ActivityScenario.launch<NewTempLanguageActivity>(intent)

        verifyPageLayout(pageCount(), pageNum)

        //when
        Espresso.onView(ViewMatchers.withId(R.id.next_button)).perform(ViewActions.click())
        Espresso.onView(ViewMatchers.withText(R.string.missing_question_answer))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed())) // dialog displayed
        Espresso.onView(ViewMatchers.withText(R.string.missing_question_answer))
            .perform(ViewActions.pressBack()) // dismiss dialog
        Espresso.onView(ViewMatchers.withText(R.string.missing_question_answer))
            .check(ViewAssertions.doesNotExist()) // dialog dismissed

        //then
        verifyPageLayout(pageCount(), pageNum)
    }
}